pkg/resources/common/validation_test.go

```diff
@@ -60,26 +67,116 @@ func TestValidateRules(t *testing.T) {
 		},
 	}
 
-	for _, testcase := range tests {
-		t.Run("global/"+testcase.name, func(t *testing.T) {
-			err := ValidateRules([]v1.PolicyRule{
-				testcase.data,
+	for _, test := range tests {
+		test := test
+		t.Run("global/"+test.name, func(t *testing.T) {
+			t.Parallel()
+
+			err := ValidateRules([]rbacv1.PolicyRule{
+				test.data,
 			}, false, gField)
-			if testcase.haserr {
+			if test.haserr {
 				require.Error(t, err)
 				return
 			}
 			require.NoError(t, err)
 		})
-		t.Run("namespaced/"+testcase.name, func(t *testing.T) {
-			err := ValidateRules([]v1.PolicyRule{
-				testcase.data,
+		t.Run("namespaced/"+test.name, func(t *testing.T) {
+			t.Parallel()
+
+			err := ValidateRules([]rbacv1.PolicyRule{
+				test.data,
 			}, true, nsField)
-			if testcase.haserr {
+			if test.haserr {
 				require.Error(t, err)
 				return
 			}
 			require.NoError(t, err)
 		})
 	}
 }
+
+func TestCheckCreatorPrincipalName(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+	userCache := fake.NewMockNonNamespacedCacheInterface[*v3.User](ctrl)
+	userCache.EXPECT().Get(gomock.Any()).DoAndReturn(func(name string) (*v3.User, error) {
```

Comment:
nit: It'd be nice to remove `gomock.Any()` and replace it with the expected parameter if possible

Reply:
This is a perfect example where you can't use anything other than `gomock.Any()` as it expects multiple values. In fact ANY value is acceptable in the context of this test.

Reply:
I don't agree with you here. `userCache.Get()` always expects the `creatorId` (`annotations[auth.CreatorIDAnn]`), and we don't verify this parameter because we are using `gomock.any()`. For example, if you change this [line](https://github.com/pmatseykanets/webhook/blob/validate-creator-principal-name/pkg/resources/common/validation.go#L79) to `user, err := userCache.Get("wrong-id")` application would stop working, but the test would still pass.

In order to get rid of `gomock.any()`, you'd need to create a mock for each test case inside the tests struct. I'd create a func that accept a string (the  `creatorId`) and returns the mock.

Something like:

```
tests := []struct {
		desc          string
		creatorID     string
		principalName string
		userCache func(string) controllerv3.UserCache
		fieldErr      bool
		err           bool
	}{

		{
			desc:          "creator id and principal name match",
			creatorID:     "u-12345",
			principalName: "keycloak_user://12345",
			userCache: func(creatorId string) controllerv3.UserCache {
				userCache := fake.NewMockNonNamespacedCacheInterface[*v3.User](ctrl)
				userCache.EXPECT().Get(creatorId).DoAndReturn(func(name string) (*v3.User, error) {
					return &v3.User{
						ObjectMeta: metav1.ObjectMeta{
							Name: "u-12345",
						},
						PrincipalIDs: []string{"local://12345", "keycloak_user://12345"},
					}, nil
				}
				return userCache
			},
		},
		{
			desc:          "error getting creator user",
			creatorID:     "u-error",
			principalName: "keycloak_user://12345",
			userCache: func(creatorId string) controllerv3.UserCache {
				userCache := fake.NewMockNonNamespacedCacheInterface[*v3.User](ctrl)
				userCache.EXPECT().Get(creatorId).DoAndReturn(func(name string) (*v3.User, error) {
					return nil, fmt.Errorf("some error")
				}
				return userCache
			},
			err:           true,
		},
	}

	for _, test := range tests {
		test := test
		t.Run(test.desc, func(t *testing.T) {
			t.Parallel()

			annotations := map[string]string{}
			if test.creatorID != "" {
				annotations[auth.CreatorIDAnn] = test.creatorID
			}
			if test.principalName != "" {
				annotations[auth.CreatorPrincipalNameAnn] = test.principalName
			}

			fieldErr, err := CheckCreatorPrincipalName(test.userCache(test.creatorID), &v3.Project{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: annotations,
				},
			})
			require.Equal(t, test.fieldErr, fieldErr != nil)
			require.Equal(t, test.err, err != nil)
		})
	}
```

You can specify different mock for each test case by using the previous approach. Then you can get rid of the [switch](https://github.com/pmatseykanets/webhook/blob/validate-creator-principal-name/pkg/resources/common/validation_test.go#L105) case that returns a different value based on the name, which can become difficult to read if this test grows in the future.

Reply:
You an find an example of this approach in https://github.com/raulcabello/rancher/blob/134a10ce985571b9856814e26011766ab0a7867f/pkg/auth/requests/impersonate_test.go#L27

Reply:
While it's possible to do this in this particular case given a relatively small number of test cases it definitely doesn't help with readability and amount of additional code for tests organized this way.

> and we don't verify this parameter because we are using 

Actually we do indirectly with the `switch`.

> For example, if you change this [line](https://github.com/pmatseykanets/webhook/blob/validate-creator-principal-name/pkg/resources/common/validation.go#L79) to user, err := userCache.Get("wrong-id") application would stop working, but the test would still pass.

Not really, no. The two main test cases break because of that indirect assertion in the `DoAndReturn()`.

That being said I think matching arguments on mock expectations is a very useful tool , especially for simple type arguments, and help to have a more comprehensive tests. But as everything else it comes with a cost.